### PR TITLE
No if subsample else

### PR DESCRIPTION
**Pull request by tmorzade**
# Description

The if_else isinstance(Subsample) are deleted for code homogenization

Fixes #(issue)

## Type of change

Please remove options that are irrelevant.
Code refactorization

# Checklist:

- [x ] I have read the [contributing guidelines](https://github.com/simai-ml/MAPIE/blob/master/CONTRIBUTING.rst)
- [ x] I have updated the [HISTORY.rst](https://github.com/simai-ml/MAPIE/blob/master/HISTORY.rst) and [AUTHORS.rst](https://github.com/simai-ml/MAPIE/blob/master/AUTHORS.rst) files
- [x] Linting passes successfully : `make lint`
- [x] Typing passes successfully : `make type-check`
- [x] Unit tests pass successfully : `make tests`
- [x] Coverage is 100% : `make coverage`
- [ ] Documentation builds successfully : `make doc`

### CODE REVIEW

File: mapie/tests/test_common.py [hunk 1]
```diff
@@ -1,19 +1,17 @@
...
 from sklearn.model_selection import KFold
-from sklearn.utils.validation import check_is_fitted
-from sklearn.exceptions import NotFittedError
+from sklearn.pipeline import make_pipeline
 from sklearn.utils.estimator_checks import parametrize_with_checks
-
-from mapie.regression import MapieRegressor
-from mapie.classification import MapieClassifier
-
+from sklearn.utils.validation import check_is_fitted
```
Comment (gmartinonQM) [inline mapie/tests/test_common.py:14]: Why changing the order of the imports ? It breaks the convention of internal packages (mapie) imported last.

If this is due to Black, please turn it off when contributing to open-source projects.

File: mapie/tests/test_regression.py [hunk 2]
```diff
@@ -2,21 +2,20 @@
...
 from sklearn.linear_model import LinearRegression
 from sklearn.model_selection import KFold, LeaveOneOut, train_test_split
 from sklearn.utils.validation import check_is_fitted
-
-from mapie.aggregation_functions import aggregate_all
-from mapie.metrics import regression_coverage_score
-from mapie.regression import MapieRegressor
-from mapie.subsample import Subsample
-from mapie._typing import ArrayLike
+from typing_extensions import TypedDict
```
Comment (gmartinonQM) [inline mapie/tests/test_regression.py:18]: Please do not change the order of the imports and turn Black off

File: mapie/tests/test_regression.py [hunk 3]
```diff
@@ -403,16 +411,17 @@ def test_invalid_aggregate_all() -> None:
         aggregate_all(None, X)
 
 
-def test_invalid_aggregate_mask() -> None:
+def test_aggregate_with_mask_with_prefit() -> None:
     """
-    Test that wrong aggregation in MAPIE raise errors.
+    Test ``aggregate_with_mask`` in case ``cv`` is "prefit".
```
Comment (gmartinonQM) [inline mapie/tests/test_regression.py:416]:
```suggestion
    Test ``aggregate_with_mask`` in case ``cv`` is ``"prefit"``.
```

File: mapie/tests/test_regression.py [hunk 4]
```diff
@@ -403,16 +411,17 @@ def test_invalid_aggregate_all() -> None:
...
     """
+
```
Comment (gmartinonQM) [inline mapie/tests/test_regression.py:418]:
```suggestion
```

**Review by gmartinonQM — CHANGES_REQUESTED** [review]
Thanks !

File: doc/theoretical_description_regression.rst [hunk 5]
```diff
@@ -178,6 +178,37 @@ methods and emphasizes their main differences.
    :width: 800
 
 
+7. The jackknife+-after-bootstrap method
+========================================
+In order to reduce the computational time, and get more robust predictions, 
+one can adopt a bootstrap approach instead of a leave-one-out approach, called 
+the jackknife+-after-bootstrap method, offered by Kim and al. [2].
```
[the remark this answers is not in the document]
Comment (tmorzade) [reply doc/theoretical_description_regression.rst:185]: I copy-paste the file from master branch on github

File: mapie/regression.py [hunk 6]
```diff
@@ -82,6 +81,16 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
 
         By default ``None``.
 
+    ensemble: bool
```
[the remark this answers is not in the document]
Comment (tmorzade) [reply mapie/regression.py:84]: Done. I also add a "check function" to ensure that agg_function is not None if ensemble is True

File: mapie/regression.py [hunk 7]
```diff
@@ -82,6 +81,16 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
...
+        Boolean determining whether the predictions, which are the first
+        item returned by the predict method, are the predictions made by
+        single_estimator_, which is trained on the whole training set, or the
+        aggregation of the predictions made by the refitted estimators, by
+        the aggregation function.
+        If cv is ``"prefit"``, ``ensemble`` is ignored.
+
+        By default ``False``.
```
[the remark this answers is not in the document]
Comment (tmorzade) [reply mapie/regression.py:92]: Replaced. But I am not sure that ensembled is the right word. Aggregated maybe?

File: mapie/regression.py [hunk 8]
```diff
@@ -95,11 +104,11 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
         By default ``None``.
 
     agg_function : str
-        Determines if and how the final prediction is aggregated from the fold
+        Determines how the final prediction is aggregated from the fold
```
[the remark this answers is not in the document]
Comment (tmorzade) [reply mapie/regression.py:107]: I do not get everything of this part. I took the last sentence

File: mapie/regression.py [hunk 9]
```diff
@@ -95,11 +104,11 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
...
         predictions.
 
-        If ``None``, returns the predictions from the single estimator
-        trained on the full training dataset.
+        If ``None``, it is ignored except if needed, in which case an error is
```
[the remark this answers is not in the document]
Comment (tmorzade) [reply mapie/regression.py:110]: if cv is Subsample. I specified it, with a Note for those setting ensemble to True in predict

File: mapie/tests/test_common.py (hunk 1 above)
Comment (tmorzade) [reply mapie/tests/test_common.py:14]: it was indeed a option in .vscode.json

File: mapie/tests/test_regression.py (hunk 2 above)
Comment (tmorzade) [reply mapie/tests/test_regression.py:18]: It is an option badly set in Black

File: mapie/tests/test_regression.py (hunk 3 above)
Comment (tmorzade) [reply mapie/tests/test_regression.py:416]: I fear I migth not get the exact suggestion

File: mapie/tests/test_regression.py (hunk 4 above)
Comment (tmorzade) [reply mapie/tests/test_regression.py:418]: Done

File: mapie/regression.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (gmartinonQM) [reply mapie/regression.py]: Yes, but it is misleading with the `agg_function` argument

File: mapie/regression.py [hunk 10]
```diff
@@ -474,29 +525,25 @@ def fit(
         # Initialization
         self.estimators_: List[RegressorMixin] = []
 
-        if isinstance(cv, Subsample):
+        # Work
+        if cv == "prefit":
+            self.single_estimator_ = estimator
+            y_pred = self.single_estimator_.predict(X)
+            self.n_samples_val_ = [X.shape[0]]
+            self.k_ = np.full(shape=(2, 2), fill_value=np.nan, dtype=float)
```
[the remark this answers is not in the document]
Comment (gmartinonQM) [reply mapie/regression.py]: nans

File: mapie/regression.py [hunk 11]
```diff
@@ -1,10 +1,11 @@
 from __future__ import annotations
 
 from typing import Iterable, List, Optional, Tuple, Union, cast
+from xmlrpc.client import Boolean
```
Comment (gmartinonQM) [inline mapie/regression.py:4]: It looks like a useless package ?

File: mapie/tests/test_regression.py (hunk 3 above)
Comment (gmartinonQM) [reply mapie/tests/test_regression.py]: Double backticks needed for sphinx rendering

**Review by gmartinonQM — CHANGES_REQUESTED** [review]
Thanks again ! Could you edit the HISTORY.rst please ?

File: mapie/regression.py [hunk 12]
```diff
@@ -109,10 +118,16 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
 
         When the cross-validation strategy is Subsample (i.e. for the
         Jackknife+-after-Bootstrap method), this function is also used to
-        aggregate the training set insample predictions. If ``None``, the
-        default aggregation function to compute residuals is then "mean".
+        aggregate the training set insample predictions.
```
[the remark this answers is not in the document]
Comment (tmorzade) [reply mapie/regression.py:115]: Done

File: mapie/regression.py [hunk 13]
```diff
@@ -109,10 +118,16 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
...
 
-        By default ``None``.
+        If cv is ``"prefit"``, ``agg_function`` is ignored.
+
+        Note: previously, when ``agg_function`` was not ``None``, the
+        prediction given by the predict method was the aggregation of the
+        predictions made by the reffited estimators. There is now an argument
+        ``ensemble`` that should be set to ``True`` for this purpose.
```
[the remark this answers is not in the document]
Comment (tmorzade) [reply mapie/regression.py:122]: Deleted

File: mapie/regression.py [hunk 14]
```diff
@@ -137,11 +152,9 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
         Residuals between ``y_train`` and ``y_pred``.
 
     k_ : ArrayLike
-        - Id of the fold containing each training sample,
-          if cv is not Resample. Of shape(n_samples_train,).
+        - [[nan, nan], [nan,nan ] if cv is ``"prefit"`` (defined but not used)
```
[the remark this answers is not in the document]
Comment (tmorzade) [reply mapie/regression.py]: ce point je ne l'ai pas compris

File: mapie/regression.py [hunk 15]
```diff
@@ -137,11 +152,9 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
...
         - Dummy array of folds containing each training sample, otherwise.
-          Of shape (n_samples_train, n_resamplings).
-
+          Of shape (n_samples_train, cv.cv.get_n_splits(X, y)).
```
[the remark this answers is not in the document]
Comment (tmorzade) [reply mapie/regression.py:152]: Done

File: mapie/regression.py [hunk 16]
```diff
@@ -196,13 +209,15 @@ def __init__(
...
+        ensemble: bool = False,
         n_jobs: Optional[int] = None,
-        agg_function: Optional[str] = None,
+        agg_function: Optional[str] = "mean",
         verbose: int = 0,
     ) -> None:
         self.estimator = estimator
         self.method = method
         self.cv = cv
+        self.ensemble = ensemble
```
[the remark this answers is not in the document]
Comment (tmorzade) [reply mapie/regression.py]: Removed

File: mapie/regression.py [hunk 17]
```diff
@@ -222,14 +237,53 @@ def _check_parameters(self) -> None:
...
 
-        if self.agg_function not in self.valid_agg_functions_:
+        check_n_jobs(self.n_jobs)
+        check_verbose(self.verbose)
+
+    def _check_agg_function(
+        self, agg_function: Optional[str] = None
+    ) -> Optional[str]:
+        """
+        Check if agg_function is correct, and coherent with other arguments.
```
[the remark this answers is not in the document]
Comment (tmorzade) [reply mapie/regression.py:240]: Done

File: mapie/regression.py [hunk 18]
```diff
@@ -222,14 +237,53 @@ def _check_parameters(self) -> None:
...
+
+        Parameters
+        ----------
+        agg_function : Optional[str], optional
+            Aggregation function's name to check, by default ``None``.
+
+        Returns
+        -------
+        str
+            ``agg_functtion`` itself or ``"mean"``.
```
[the remark this answers is not in the document]
Comment (tmorzade) [reply mapie/regression.py:250]: Corrected

File: mapie/regression.py [hunk 19]
```diff
@@ -222,14 +237,53 @@ def _check_parameters(self) -> None:
...
+
+        Raises
+        ------
+        ValueError
+            If ``agg_function`` is not in [``None``, "mean", "median"], or is
```
[the remark this answers is not in the document]
Comment (tmorzade) [reply mapie/regression.py:255]: Corrected

File: mapie/regression.py [hunk 20]
```diff
@@ -222,14 +237,53 @@ def _check_parameters(self) -> None:
...
+            ``None`` while needed.
+
```
[the remark this answers is not in the document]
Comment (tmorzade) [reply mapie/regression.py:257]: Deleted

File: mapie/regression.py [hunk 21]
```diff
@@ -222,14 +237,53 @@ def _check_parameters(self) -> None:
...
+        """
+
```
[the remark this answers is not in the document]
Comment (tmorzade) [reply mapie/regression.py:259]: Quelle différence avec la remarque précédente

File: mapie/regression.py [hunk 22]
```diff
@@ -415,15 +462,18 @@ def aggregate_with_mask(self, x: ArrayLike, k: ArrayLike) -> ArrayLike:
...
-            # with phi2D(A=x, B=k, fun=lambda x: np.nanmean(x, axis=1).
-            # However, phi2D contains a np.apply_along_axis loop which
-            # is much slower than the matrices multiplication that can
-            # be used to compute the means.
-            K = np.nan_to_num(k, nan=0.0)
-            return np.matmul(x, (K / (K.sum(axis=1, keepdims=True))).T)
-        raise ValueError("Aggregation function called but not defined.")
+        if self.cv == "prefit":
+            raise ValueError(
+                "There should not have aggragtion of prediction if cv is "
```
[the remark this answers is not in the document]
Comment (tmorzade) [reply mapie/regression.py:479]: Corrected

File: mapie/regression.py [hunk 23]
```diff
@@ -415,15 +462,18 @@ def aggregate_with_mask(self, x: ArrayLike, k: ArrayLike) -> ArrayLike:
...
+                "'prefit'"
+            )
+        # To aggregate with mean() the aggregation coud be done
+        # with phi2D(A=x, B=k, fun=lambda x: np.nanmean(x, axis=1).
+        # However, phi2D contains a np.apply_along_axis loop which
+        # is much slower than the matrices multiplication that can
+        # be used to compute the means.
+        K = np.nan_to_num(k, nan=0.0)
+        return np.matmul(x, (K / (K.sum(axis=1, keepdims=True))).T)
```
[the remark this answers is not in the document]
Comment (tmorzade) [reply mapie/regression.py:488]: Non, parce que si agg_function is None, par défaut je mets "mean" mais sans changer self.agg_function

File: mapie/regression.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (tmorzade) [reply mapie/regression.py]: Done

File: mapie/tests/test_regression.py (hunk 3 above)
Comment (tmorzade) [reply mapie/tests/test_regression.py]: Done, I think

File: mapie/regression.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (gmartinonQM) [reply mapie/regression.py]: It is just a typo: white space after the come, not before the hook. Is it clearer now ?

File: mapie/regression.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (gmartinonQM) [reply mapie/regression.py:253]: No blank line after a docstring

File: mapie/regression.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (gmartinonQM) [reply mapie/regression.py]: But the comment is outdated since you changed the docstring

File: mapie/regression.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (gmartinonQM) [reply mapie/regression.py:482]: Right, but here, this code is executed even if `agg_function` is not "mean", provided it is not "median" either. Furthermore, it is not clear from the code that `None` is equivalent to `"mean"`. Could you be more explicit with for example a 

```
if self.agg_function in ["mean", None]:

**Review by gmartinonQM — CHANGES_REQUESTED** [review]
Thanks ! Some comments still dangling out

File: mapie/regression.py (hunk 11 above)
Comment (tmorzade) [reply mapie/regression.py:4]: Indeed, I delete it

File: mapie/regression.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (tmorzade) [reply mapie/regression.py:482]: Done

[the remark this answers is not in the document]
Comment (tmorzade) [reply mapie/regression.py:482]: with an extended  corresponding test

File: mapie/regression.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (tmorzade) [reply mapie/regression.py:253]: Done

**Review by gmartinonQM — COMMENTED** [review]
LGTM !

File: mapie/regression.py [hunk 24]
```diff
@@ -222,14 +223,46 @@ def _check_parameters(self) -> None:
...
+        """
+        if agg_function not in self.valid_agg_functions_:
             raise ValueError(
                 "Invalid aggregation function "
                 "Allowed values are None, 'mean', 'median'."
             )
 
-        check_n_jobs(self.n_jobs)
-        check_verbose(self.verbose)
+        if isinstance(self.cv, Subsample) and (agg_function is None):
```
Comment (vtaquet) [inline mapie/regression.py:258]: I'm seeing this a bit too late, but why does the user now have to specify explicitely an `agg_function` now ? If I'm correct, when `agg_function` was not specified, a "mean" was used, why this change ?

**Review by vtaquet — COMMENTED** [review]
Looks good to me ! I just have a minor question.
